src/ray/object_manager/object_buffer_pool.h

```diff
@@ -0,0 +1,186 @@
+  /// Returns a chunk of an empty object at the given chunk_index. The object chunk
+  /// serves as the data that is to be written to by a connection receiving an object
+  /// from a remote node.
+  ///
+  /// \param object_id The ObjectID.
+  /// \param data_size The sum of the object size and metadata size.
+  /// \param metadata_size The size of the metadata.
+  /// \param chunk_index The index of the chunk.
+  /// \return A ChunkInfo struct.
+  const ChunkInfo &CreateChunk(const ObjectID &object_id, uint64_t data_size,
+                               uint64_t metadata_size, uint64_t chunk_index);
+
+  /// When a chunk is done being used as part of a create, this method is invoked
+  /// to indicate that the buffer associated with a chunk is either ready to be
+  /// sealed or aborted. Creation is aborted if the call to plasma client Create failed.
+  ///
+  /// \param object_id The ObjectID.
+  /// \param succeeded Whether the operation on the chunk succeeded. If it failed,
+  /// object creation will be aborted.
+  /// \return The status of invoking this method.
+  ray::Status SealOrAbortBuffer(const ObjectID &object_id, bool succeeded);
+
+  /// Terminates this object.
+  /// This must be called after remote client connections have been terminated.
+  void Terminate();
+
+ private:
+  /// Seal the object associated with a create operation.
+  ray::Status SealCreate(const ObjectID &object_id);
+
+  /// Abort the create operation associated with an object.
+  ray::Status AbortCreate(const ObjectID &object_id);
+
+  /// Abort the get operation associated with an object.
+  ray::Status AbortGet(const ObjectID &object_id);
+
+  /// Builds the chunk vector for an object, and store it by object_id in chunk_info_.
+  /// Returns the number of chunks into which object is split.
+  uint64_t BuildChunks(const ObjectID &object_id, uint8_t *data, uint64_t data_size,
+                       uint64_t metadata_size);
+
+  /// Get an object from the object store pool.
+  std::shared_ptr<plasma::PlasmaClient> GetObjectStore();
+
+  /// Release an object from the object store pool.
+  void ReleaseObjectStore(std::shared_ptr<plasma::PlasmaClient> client);
+
+  /// Adds a client to the client pool and mark it as available.
+  void Add();
+
+  /// The type of a buffer state.
+  /// Used to detect whether an object buffer is
+  /// both a get and create buffer.
+  enum class BufferStateType : int { GET = 0, CREATE };
+
+  /// Holds the state of a buffer.
+  struct BufferState {
+    BufferState() {}
+    BufferState(std::shared_ptr<plasma::PlasmaClient> client, uint64_t references,
+                BufferStateType state_type)
+        : client(client), references(references), state_type(state_type) {}
+    /// The plasma client used by this buffer.
+    std::shared_ptr<plasma::PlasmaClient> client;
+    /// The number of references that currently rely on this buffer.
+    /// We expect this many calls to Release or SealOrAbortBuffer.
+    uint64_t references;
```

Comment:
An alternative design that comes to mind is to combine the `chunk_info_` and `buffer_state_` maps. Also, instead of storing both a `std::vector` of the chunk information and a `references` count, how about we just make it a `std::list` of the chunk information, and the reference count is just the size of the list?